Progress cli support

mhdawson opened the pull request:
- fix up some error checking when command are not included or invalid command is provided
- start filling in show command
- add scaffolding for testing ci commands

This depends on https://github.com/pkgjs/support/pull/1 landing first

mhdawson committed:
fix: make sure we get proper help output

Help output was incomplete when no command was
specified and was not printed when an invalid
command was specified.

Signed-off-by: Michael Dawson <michael_dawson@ca.ibm.com>

mhdawson committed:
feat: start filling in show command

- move support info out of package.json
- code to walk/print tree of modules
- code to show local support info if available

Signed-off-by: Michael Dawson <michael_dawson@ca.ibm.com>

mhdawson committed:
feat add option so prefer canonical support info

Add --canonical option to prefer canonical support info
even when support info for a module is available locally

Signed-off-by: Michael Dawson <michael_dawson@ca.ibm.com>

mhdawson committed:
feat: expand coverage of support specification

- expand coverage of options for how to specify support
  file location
- add initial cli based test suite

Signed-off-by: Michael Dawson <michael_dawson@ca.ibm.com>

mhdawson commented:
@wesleytodd , @Eomm wanted to get this in to make sure we don't duplicate efforts. There was some discussion around whether we should land in master incrementally or work on the cli branch instead. 

I'd vote for incrementally landing on master as I think it will potentially be more visible and I don't see a downside since the the project is just ramping up.

Eomm commented:
I agree to squash&merge to give to other people a baseline and a pattern to starts with other issues: a branch is not so viewed 👍

mhdawson committed:
fix: refactor to share common code

- move common code up into support object
- use commond code from show and validate comamnds
- use validate code in support object validate cli command
- add initial validate cli tests

Signed-off-by: Michael Dawson <michael_dawson@ca.ibm.com>

mhdawson commented:
Pushed another commit to refactor to share code across commands and add some cli testing for validate (much more is needed)

mhdawson committed:
fix: simplify argument handling

  - remove unneeded overhead in arg handling

  Signed-off-by: Michael Dawson <michael_dawson@ca.ibm.com>

mhdawson committed:
feat: add support for --canonical to validate

- add support for --canonical to validate
- add test for non-resolved support info to validate

Signed-off-by: Michael Dawson <michael_dawson@ca.ibm.com>

wesleytodd commented:
Oh I rebased and force pushed on that other branch.  I also fixed @ljharb's issue that it had a path escape issue.  Should probably rebase this on top of that before merging as well!

mhdawson commented:
@wesleytodd how long ago did  you do that? I rebased earlier today

mhdawson commented:
And it seems to say there are no conflicts.

wesleytodd commented:
Yesterday!  I had that comment typed in an open tab, looks like your changes did not cause it to update the ui.  I just hit "comment" with the text I typed yesterday lol.

mhdawson committed:
feat: add option to fetch support information

- add option to fetch support information from
  github repository when requested.

Signed-off-by: Michael Dawson <michael_dawson@ca.ibm.com>

wesleytodd commented:
When we are ready to merge this, we should be sure to squash this and add all the co-author info.  https://github.com/pkgjs/support/pull/1#issuecomment-628237049

mhdawson committed:
feat: add validation of support element

- Add validation of support element. Previous
  validation was only for package-support.json
  and did not include validation of the support
  element in Package.json

- Fix validation support to work on more than just
  'true' for support

- Add a number of validation tests

Signed-off-by: Michael Dawson <michael_dawson@ca.ibm.com>

mhdawson committed:
feat: add remote fetch for validation caommand

- enable remote fetch for validation command
- add test of validation after remote fetch

Signed-off-by: Michael Dawson <michael_dawson@ca.ibm.com>

mhdawson commented:
Pushed commits to add validation of the support element in package.json and to support remote fetch for validation.

ljharb commented on index.js:
```diff
@@ -1,9 +1,35 @@
 'use strict'
+const path = require('path')
```
lol i hope we can have a semicolons discussion at some point (not in this PR)

mhdawson replied:
@ljharb not sure what you mean by the semicolons discussion but the current linter config prohibits their use in most places  (which I'll admit I'm still getting used to)

ljharb replied:
Right, I’m saying I’d like to discuss changing that (very nonstandard compared to the ecosystem) linter config, but i don’t want to derail PRs with it :-)

mhdawson replied:
ok, makes sense.

wesleytodd replied:
Can we have this conversation in slack?  I also want semi colons, but the work it takes to maintain a standard fork is non-trivial.  I had a short chat with Feross on twitter about this, and I think we could automate a fork. I just haven't had time to dig in.

Edit: I should also add `tabs`. The fact that despite the accessibility issues with spaces, we still use them by default is becoming more and more clearly the wrong decision.

ljharb replied:
Yes, let's chat about it in slack. (The solution here is to use `eslint --fix`, so we match the majority of the ecosystem and also aren't tied to unidiomatic configuration choices)

wesleytodd replied:
We chatted in slack. The conclusion is that I will do whatever the rest of the group agrees. Closing this for now.

mhdawson committed:
fix: fix show tree display

- fix show tree display which was broken by earlier
  refactoring to make functions async
- add test to validate multilevel tree

Signed-off-by: Michael Dawson <michael_dawson@ca.ibm.com>

mhdawson committed:
feat: flatten default support info out

- flatten default support info output so that
  it fits on one line. This allows the package
  hierarchy to be seen in the show command

Signed-off-by: Michael Dawson <michael_dawson@ca.ibm.com>

mhdawson committed:
fix: fixup after rebase on main branch

mhdawson commented:
Rebased on https://github.com/pkgjs/support/pull/1.  

Next step is to look at modules for prettying up the ajv validation output.

mhdawson committed:
feat: improve validation errors

Signed-off-by: Michael Dawson <michael_dawson@ca.ibm.com>

mhdawson committed:
fix: add test for invalid package-support.json

Signed-off-by: Michael Dawson <michael_dawson@ca.ibm.com>

mhdawson committed:
feat: provide API compatible errors for API

- instead of using cli based error format for API calls, use
  js format unless cli formatting is requested

Signed-off-by: Michael Dawson <michael_dawson@ca.ibm.com>

mhdawson committed:
feat: tighten up schema

- Tighten up schema
- Add tests to validate tigher schema

Signed-off-by: Michael Dawson <michael_dawson@ca.ibm.com>

mhdawson committed:
feat: make output consistent

Signed-off-by: Michael Dawson <michael_dawson@ca.ibm.com>

mhdawson commented:
@ljharb , @wesleytodd I believe I've integrated the changes we discussed when we last got together to discuss progress so far.   I think this should satisfy the MVP and is worth landing before we do a final review to see if we are ready to use it as the MVP for promoting the support info as a best practice.

wesleytodd commented on lib/cli/commands/show.js:
```diff
@@ -7,7 +35,13 @@ module.exports = function (opts) {
     },
     handler: function (argv) {
-      argv.log.info('show')
+      const Arborist = require('@npmcli/arborist')
```
I personally would prefer we not `require` deeply in code. While there might be some benefits sometimes, IMO they are all outweighed by the unexpected behavior of having a require fail only when running the program in a specific way.  If we hit a performance bottleneck where this require statement meaningfully slows down things, then worth considering IMO, but not until we do hit that.

mhdawson replied:
agreed will fix

wesleytodd commented on lib/cli/commands/validate.js:
```diff
@@ -16,47 +15,54 @@ module.exports = function (opts) {
+        if ((supportData !== true) && !(supportData instanceof Object)) {
+          console.log(errorInSupportElement)
+          console.log('Support element should either be true or an Object matching the required schema')
+          return
```
nitpick: `else` is redundant with early returns, prefer reduced nesting where possible.

mhdawson replied:
will fix

wesleytodd commented on test/artifacts/MySupportInfo.json:
```diff
@@ -0,0 +1,16 @@
+{
```
Totally not a blocker, but this file naming is surprising to me in two ways, just wondering your thoughts:

1. I typically see this kind of thing called a `fixture`, is the convention of `artifacts` from another school of thought?
2. All the other files are kebab, this is camel.  While this might not bite anyone on a regular basis, avoiding upper case in file names means you don't worry about case-insensitive file systems.  Here is both goes counter to the other files and has uppercase in the filename.  Is there a reason?

mhdawson replied:
will fix

wesleytodd commented on test/cli/validate-local-bad-element3/command:
```diff
@@ -0,0 +1 @@
+validate 
```
Maybe I haven't gotten to where these are used yet, but what is the point of all these files with just the string `validate` in them?

wesleytodd commented on test/index.js:
```diff
@@ -40,5 +42,21 @@ const support = require('..')
+        assert.strictEqual(result, expected)
+      })
+    })
+  })
```
I think I see it together now, but I think disagree with the approach of making the code abstract and having duplicated files. If the abstraction did end up reducing duplication or the code complexity, then I could see adding all this but IMO this just made the PR much more difficult to review.

TBQH, I might prefer to see one object with all of this in the test file and the abstraction on the bottom than having a bunch of files with just the string `validate` in them.  IDK, I am sure this is a matter of opinion and others probably disagree with me 😛.  I am also sure there is a middle ground that would also put me more at ease with this approach.

mhdawson replied:
Each test has a file for

- command to run
- expected output
- other required content like package.json package-support.json  required for the test.

In this case the file with `validate` is the command to be run for the test, in some cases there are additional options so they are not all the same.

So your suggestion is one file with multiple sections versus multiple files?  That will be more work to manage since right now its easy to generate the output file by just running the command.  Merging them into 1 will require having to merge, dealing with any required escaping etc.

wesleytodd replied:
I think what really thew me was the `command` files.  I was just scrolling and scrolling this PR and seeing the same file repeated.  I do understand having the content as fixtures in a directory, as that is really a requirement.  I don't think this would be something I would ever block, but it all seems more difficult to grok to me. 🤷

mhdawson replied:
I'll do a pass through all of the other comments and fix up and then we can loop back to this one. I'd prefer to not block on this but we can update/change the tests to use a different structure if we find a better way in a later PR.

helio-frota replied:
my comment was only a reminder not a blocker btw :+1:

wesleytodd replied:
Agreed, this is not a blocker.  Once we get this all out and in use we can just open an issue to get on the same page for test strategy.

helio-frota commented on lib/cli/index.js:
```diff
@@ -56,6 +56,21 @@ module.exports = (options = {}) => {
       'log-level': ['silent', 'verbose']
     })
+    .option('canonical', {
+      describe: 'Use canaonical support info even when local info is available',
```
nitpick: minor typo `canaonical`

wesleytodd commented:
One other comment, the commit messages, can we adopt [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/)?  If we do, then I can go ahead with another effort I planned, which was to integrate either [Release Please](https://github.com/googleapis/release-please) or at least an auto generated change log and version script like [`standard-version`](https://www.npmjs.com/package/standard-version).

I think the change here would be to `git rebase -i` and edit each commit to have `fix` or `feat` in it.

ljharb commented:
i'm super opposed to conventional commits spec, because i find the use of the word "chore" to be hugely toxic - i'm fine with the concept (tooling to enforce commit message conventions), but afaik the tooling isn't customizable in that way?

wesleytodd commented:
Interesting, I haven't thought through that lens. I agree that `chore` is not great, but that doesn't technically come from the conventional commit spec, that is an angular thing.  I am sure we could get changes to that terminology, would `release` do?

mhdawson committed:
fix: add missing test files

- Needed to force including files that matched
  git exclude

mhdawson committed:
fix: updated expected output based on arborist fix

Signed-off-by: Michael Dawson <michael_dawson@ca.ibm.com>

mhdawson committed:
fix: address review comments

Signed-off-by: Michael Dawson <michael_dawson@ca.ibm.com>

ljharb commented:
@wesleytodd it's usually used for updating dependencies, fixing config files - things that aren't tied to a release or test/impl contents, but are still critically important for operation of the repo. My personal convention uses `[Deps]`, `[Dev Deps]`, `[Tests]`, `[Meta]`, etc, rather than lumping them all under one category.

wesleytodd commented:
Cool, so I am not super opinionated on what sub categories we use, the spec leaves that open ended.  So examples from yours above would be:

```
fix(tests): test were broken
fix(meta): updated docs
feat(deps): updated foo@1.2.3 adds support for feature X
fix(dev deps): updated bar@1.2.3 better perf for tests
fix(deps)!: updated foo@2.0.0 breaking change, we expose foo's api to end users
```

mhdawson commented:
@wesleytodd I've pushed a commit to address all of your comments except for the commit comments.

I tried the rebase but it complained about problems being able to apply commits, possibly due to changes pushed after my last change (seems like you might have forced pushed?).  Since Github says the PR can be landed I suggest we do that and then I can re-checkout and update the commit comments and the force push to update.

If you agree I'll go ahead and to that.

@helio-frota I've also fixed up your comment so I'll dismiss your request changes.

mhdawson commented:
actually seems that if I try to merge without squashing if does say there are problems :(

mhdawson committed:
fix: fixup after rebase on master

- fixup after manual cherry-pick of commits to be
  able to rebase on master

Signed-off-by: Michael Dawson <michael_dawson@ca.ibm.com>

mhdawson commented:
Ok took a bit of work cherry-picking manually but have fixed it up.

@helio-frota, @wesleytodd all of your comments should be addressed.

wesleytodd commented on lib/cli/index.js:
```diff
@@ -56,6 +56,21 @@ module.exports = (options = {}) => {
       silent: 'verbose',
       'log-level': ['silent', 'verbose']
     })
+    .option('canonical', {
```
Just a question, this is the equivalent meaning of "--prefer-online" right?  So always fetch?

wesleytodd replied:
If the above, then what is the difference with `fetch`?   Maybe the subtlety in the difference is confusing.  If it is for me, I guess it will also be for end users.

mhdawson replied:
Canonical means use the "canonical" source of truth which is the online version. If you specify that without "fetch" it will show you the URL for the canonical data but won't fetch it unless fetch is specified.  This lets you get info about where the canonical data is without being forced to fetch it.

Once we land this PR my next step will be a PR for doc to provide more info on how to use the commands.

wesleytodd replied:
Oh right!! I remember now.  I do think this behavior will need good docs because it is a bit confusing.  Thanks for the clarification.

wesleytodd commented on lib/cli/index.js:
```diff
@@ -56,6 +56,21 @@ module.exports = (options = {}) => {
+      type: 'boolean',
+      group: 'preferences'
+    })
+    .option('base-path', {
```
This is intended to be the root of the repo correct?  So you can load support info up the tree?  Just want to make sure I understand.

If so, we might also want to come up with a better `describe` block.  Maybe `Allow support.json files to be loaded from under this base path (default: cwd)`?  Even this is not great, I am just not sure how to describe it well to an end user.

mhdawson replied:
@wesleytodd 

This is the root of where you want to allow local support info files to be loaded from.

By default you can't load a file outside in directories above the directory with the package.json. For a since module repo that may be ok, but for monorepos or packages that share a common support-info.json you need to set this if you want to load them locally.

Maybe "Set top most directory from which local support info files can be read. Defaults to same directory in which package.json is located."

wesleytodd replied:
I like that description!  I think I understood it correctly, but the new wording makes it much more clear I think.

wesleytodd commented on package.json:
```diff
@@ -3,12 +3,14 @@
     "sustainability"
   ],
   "license": "MIT",
+  "support": true,
```
duplicated key

mhdawson replied:
will fix

mhdawson committed:
fix: address additional review comments

Signed-off-by: Michael Dawson <michael_dawson@ca.ibm.com>

mhdawson commented:
@wesleytodd pushed additional commit to address new comments.

wesleytodd reviewed (approved):
I think only the last to comments are yet to resolve.  Do you plan on keeping the commits or squashing this down to one?  I am not strongly opinionated either way, so LGTM!

EDIT: posted at the same time you pushed.

mhdawson commented:
I was thinking of leaving the commits as they show the build up of the commands/functionality/testing so the history might be interesting. (Hopefully for subsequent commits we can get code into the repo in smaller chunks and squashing would make sense)

mhdawson commented:
Going to land. I believe I addressed @helio-frota comment and he mentioned it was not blocking in the discussion.
